packages/dataviews/src/dataform-controls/array.tsx

```diff
@@ -18,68 +22,183 @@ export default function ArrayControl< Item >( {
+				if ( typeof token === 'object' && 'value' in token ) {
 					return token.value;
 				}
+				return token;
+			} );
+
+			// First, check if elements validation is required and any tokens are invalid
+			if ( field.isValid?.elements && elements ) {
+				const invalidTokens = tokenValues.filter( ( tokenValue ) => {
+					return ! elements.some(
+						( element ) => element.value === tokenValue
+					);
+				} );
+
+				if ( invalidTokens.length > 0 ) {
+					setCustomValidity( {
+						type: 'invalid',
+						message: __(
+							'Please select from the available options.'
+						),
+					} );
+					return;
+				}
+			}
+
+			// Then check custom validation if provided
+			if ( field.isValid?.custom ) {
+				const result = field.isValid.custom(
+					{
+						...data,
+						[ id ]: tokenValues,
+					},
+					field
+				);
+
+				if ( result ) {
+					setCustomValidity( {
+						type: 'invalid',
+						message: result,
+					} );
+					return;
+				}
+			}
 
-				const tokenByLabel = findElementByLabel( token );
+			// If no validation errors, clear custom validity
+			setCustomValidity( undefined );
+		},
+		[ elements, data, id, field ]
+	);
 
-				return tokenByLabel?.value || token;
+	// Generate custom invalid message for elements validation
+	const customInvalidMessage = useMemo( () => {
+		if ( field.isValid?.elements && elements ) {
+			return __( 'Please select from the available options.' );
+		}
+		return __( 'Invalid item' );
+	}, [ field.isValid?.elements, elements ] );
+
+	const onChangeControl = useCallback(
+		( tokens: ( string | { value: string; label?: string } )[] ) => {
+			const valueTokens = tokens.map( ( token ) => {
+				if ( typeof token === 'object' && 'value' in token ) {
+					return token.value;
+				}
+				// If it's a string, it's either a new suggestion value or user input
+				return token;
 			} );
 
 			onChange( {
-				[ id ]: stringTokens,
+				[ id ]: valueTokens,
 			} );
+
+			validateTokens( tokens );
 		},
-		[ id, onChange, findElementByLabel ]
+		[ id, onChange, validateTokens ]
 	);
 
+	const onInputChange = useCallback( ( input: string ) => {
+		if ( input === '' ) {
+			// Reset custom validity when input is cleared
+			setCustomValidity( undefined );
+		}
+	}, [] );
+
 	return (
-		<FormTokenField
+		<ValidatedFormTokenField
+			required={ !! field.isValid?.required }
+			onValidate={ validateTokens }
+			customValidity={ customValidity }
 			label={ hideLabelFromVision ? undefined : label }
-			value={ arrayValue }
+			value={ arrayValueAsElements }
 			onChange={ onChangeControl }
+			onInputChange={ onInputChange }
 			placeholder={ placeholder }
-			suggestions={
-				elements?.map( ( suggestion ) => suggestion.label ) ?? []
-			}
+			suggestions={ elements?.map( ( element ) => element.value ) }
+			messages={ {
+				added: __( 'Item added.' ),
+				removed: __( 'Item removed.' ),
+				remove: __( 'Remove item' ),
+				__experimentalInvalid: customInvalidMessage,
+			} }
```

Comment:
The added/removed/remove messages are the same as the ones provided by the control by default. The __experimentalInvalid is the same as the one provided by the `customValidity`. So I wonder if this is ever used?

This is what I did to test:

- Set value in this component to `value={ [ { value: 'howdy', label: 'Howdy' } ] }` so there's an invalid value not part of the elements.
- Changed the `customInvalidMessage` to `MESSAGE`.
- Tested the story, but I was only able to get the `customValidity` message:

<img width="385" height="315" alt="Screenshot 2025-09-17 at 11 41 11" src="https://github.com/user-attachments/assets/177e2b40-4a6a-4b36-ad22-5ef96ba2619c" />